worker/upgradedatabase/worker.go

```diff
@@ -0,0 +1,267 @@
+	upgradeComplete gate.Lock
+
+	tag            names.Tag
+	agent          agent.Agent
+	logger         Logger
+	pool           Pool
+	performUpgrade func(version.Number, []upgrades.Target, func() upgrades.Context) error
+	retryStrategy  utils.AttemptStrategy
+
+	fromVersion version.Number
+	toVersion   version.Number
+}
+
+// NewWorker validates the input configuration, then uses it to create,
+// start and return an upgradeDB worker.
+func NewWorker(cfg Config) (worker.Worker, error) {
+	var err error
+
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &upgradeDB{
+		upgradeComplete: cfg.UpgradeComplete,
+		tag:             cfg.Tag,
+		agent:           cfg.Agent,
+		logger:          cfg.Logger,
+		performUpgrade:  cfg.PerformUpgrade,
+		retryStrategy:   cfg.RetryStrategy,
+	}
+	if w.pool, err = cfg.OpenState(); err != nil {
+		return nil, err
+	}
+
+	w.tomb.Go(w.run)
+	return w, nil
+}
+
+func (w *upgradeDB) run() error {
+	defer func() {
+		if err := w.pool.Close(); err != nil {
+			w.logger.Errorf("failed closing state pool: %v", err)
+		}
+	}()
+
+	if w.upgradeDone() {
+		return nil
+	}
+
+	isPrimary, err := w.pool.IsPrimary(w.tag.Id())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// If we are the primary we need to run the upgrade steps.
+	// Otherwise we watch state and unlock once the primary has run the steps.
+	if isPrimary {
+		w.runUpgrade()
+	} else {
+		w.watchUpgrade()
+	}
+
+	return nil
+}
+
+// upgradeDone returns true if this worker
+// does not need to run any upgrade logic.
+func (w *upgradeDB) upgradeDone() bool {
+	// If we are already unlocked, there is nothing to do.
+	if w.upgradeComplete.IsUnlocked() {
+		return true
+	}
+
+	// If we are already on the current version, there is nothing to do.
+	w.fromVersion = w.agent.CurrentConfig().UpgradedToVersion()
+	w.toVersion = jujuversion.Current
+	if w.fromVersion == w.toVersion {
+		w.logger.Debugf("database upgrade to %v already completed", w.toVersion)
+		w.upgradeComplete.Unlock()
+		return true
+	}
+
+	return false
+}
+
+func (w *upgradeDB) runUpgrade() {
+	w.setStatus(status.Started, fmt.Sprintf("upgrading database to %v", w.toVersion))
+
+	if err := w.agent.ChangeConfig(w.runUpgradeSteps); err == nil {
+		w.logger.Infof("database upgrade to %v completed successfully.", w.toVersion)
+		w.setStatus(status.Started, fmt.Sprintf("database upgrade to %v completed", w.toVersion))
+		w.upgradeComplete.Unlock()
+	}
+}
+
+// runUpgradeSteps runs the required database upgrade steps for the agent,
+// retrying on failure.
+func (w *upgradeDB) runUpgradeSteps(agentConfig agent.ConfigSetter) error {
+	var upgradeErr error
+	contextGetter := w.contextGetter(agentConfig)
+
+	for attempt := w.retryStrategy.Start(); attempt.Next(); {
+		upgradeErr = w.performUpgrade(w.fromVersion, []upgrades.Target{upgrades.DatabaseMaster}, contextGetter)
+		if upgradeErr != nil {
+			w.reportUpgradeFailure(upgradeErr, attempt.HasNext())
+		}
```

Comment:
This is the bug.
```
} else {
  break
}
```

Essentially, the issue is that runUpgradeSteps never returns because the runUpgradeSteps retryStrategy gets stuck in a loop, and I think each performUpgrade after the first says "I'm already there" so it just spins. And RetryStrategy is:
				RetryStrategy:   utils.AttemptStrategy{Delay: 2 * time.Minute, Min: 5},

There is no Total, just a Min, which means we should try at least 5 times and delay at least 2 minutes each. So it retries for... more than 10 minutes.